pkg/streamer/reader_test.go

```diff
@@ -0,0 +1,689 @@
+			break
+		}
+	}
+	c.Assert(fakeRotateEventCount, Equals, 0)
+	t.verifyNoEventsInStreamer(c, s)
+
+	// generate another non-fake RotateEvent
+	rotateEv, err := event.GenRotateEvent(baseEvents[0].Header, uint32(latestPos), []byte("mysql-bin.888888"), 4)
+	c.Assert(err, IsNil)
+	_, err = f.Write(rotateEv.RawData)
+	c.Assert(err, IsNil)
+
+	// latest is still the end_log_pos of the last event, not the next relay file log file's position
+	needSwitch, needReParse, latestPos, nextUUID, nextBinlogName, err = r.parseFile(
+		ctx, s, filename, offset, relayDir, firstParse, currentUUID, possibleLast)
+	c.Assert(err, IsNil)
+	c.Assert(needSwitch, IsFalse)
+	c.Assert(needReParse, IsFalse)
+	c.Assert(latestPos, Equals, int64(rotateEv.Header.LogPos))
+	c.Assert(nextUUID, Equals, "")
+	c.Assert(nextBinlogName, Equals, "")
+	t.purgeStreamer(c, s)
+
+	// parse from a non-zero offset
+	offset = int64(rotateEv.Header.LogPos - rotateEv.Header.EventSize)
+	needSwitch, needReParse, latestPos, nextUUID, nextBinlogName, err = r.parseFile(
+		ctx, s, filename, offset, relayDir, firstParse, currentUUID, possibleLast)
+	c.Assert(err, IsNil)
+	c.Assert(needSwitch, IsFalse)
+	c.Assert(needReParse, IsFalse)
+	c.Assert(latestPos, Equals, int64(rotateEv.Header.LogPos))
+	c.Assert(nextUUID, Equals, "")
+	c.Assert(nextBinlogName, Equals, "")
+
+	// should only got a RotateEvent and a FormatDescriptionEven
+	i = 0
+	for {
+		ev, err2 := s.GetEvent(ctx)
+		c.Assert(err2, IsNil)
+		switch ev.Header.EventType {
+		case replication.ROTATE_EVENT:
+			c.Assert(ev.RawData, DeepEquals, rotateEv.RawData)
+			i++
+		case replication.FORMAT_DESCRIPTION_EVENT:
+			i++
+		default:
+			c.Fatalf("got unexpected event %+v", ev.Header)
+		}
+		if i >= 2 {
+			break
+		}
+	}
+	t.verifyNoEventsInStreamer(c, s)
+}
+
+func (t *testReaderSuite) TestParseFileRelaySubDirUpdated(c *C) {
+	var (
+		filename     = "test-mysql-bin.000001"
+		nextFilename = "test-mysql-bin.000002"
+		baseDir      = c.MkDir()
+		offset       int64
+		firstParse   = true
+		possibleLast = true
+		baseEvents   = t.genBinlogEvents(c, 0)
+		currentUUID  = "b60868af-5a6f-11e9-9ea3-0242ac160006.000001"
+		relayDir     = filepath.Join(baseDir, currentUUID)
+		fullPath     = filepath.Join(relayDir, filename)
+		nextPath     = filepath.Join(relayDir, nextFilename)
+		s            = newLocalStreamer()
+		cfg          = &BinlogReaderConfig{RelayDir: baseDir}
+		r            = NewBinlogReader(cfg)
+	)
+
+	// create the current relay log file and write some events
+	err := os.MkdirAll(relayDir, 0744)
+	c.Assert(err, IsNil)
+	f, err := os.OpenFile(fullPath, os.O_CREATE|os.O_WRONLY, 0644)
+	c.Assert(err, IsNil)
+	defer f.Close()
+	_, err = f.Write(replication.BinLogFileHeader)
+	c.Assert(err, IsNil)
+	for _, ev := range baseEvents {
+		_, err = f.Write(ev.RawData)
+		c.Assert(err, IsNil)
+	}
+
+	// no valid update for relay sub dir, timeout
+	ctx1, cancel1 := context.WithTimeout(context.Background(), time.Second)
```

Comment:
it is safe to set a short timeout value here

Reply:
done in https://github.com/pingcap/dm/pull/140/commits/ec91143d6bb3dea85021f1f3384bff2a51a02148